include/boost/gil/io/path_spec.hpp

```diff
@@ -41,9 +42,11 @@ inline std::string convert_to_string( std::string const& obj)
 
 inline std::string convert_to_string( std::wstring const& s )
 {
-    std::size_t len = wcslen( s.c_str() );
-    char* c = reinterpret_cast<char*>( alloca( len ));
-    wcstombs( c, s.c_str(), len );
+    std::mbstate_t state = std::mbstate_t();
+    const wchar_t* str = s.c_str();
+    const std::size_t len = std::wcsrtombs(nullptr, &str, 0, &state);
+    char* c = reinterpret_cast<char*>( alloca( len + 1));
```

Comment:
Is `alloca` really needed here?

Reply:
I don't really have an opinion about it. What would you use instead? `alloca` was used in the original code and I just kept it.

The thing is, we need some kind of `char*` for `wcstombs()` to write to, and `alloca` just allocates the temporary buffer here and frees it once the conversion is done. It's kind of convenient, because there is no need to free the buffer explicitly - so there is no chance to leak the buffer.

Reply:
There's already a buffer in form of `std::string`.

Reply:
No, not (yet) at this point. The `std::string` is only created three lines further down.

So let's say we decide to move that up and create a string with the proper length before the call to `wcstombs()`. Then we would still need to call `c_str()` on that string to get the "buffer" and then write to that. Looks like a possible solution, but [cppreference says](https://en.cppreference.com/w/cpp/string/basic_string/c_str):

> Writing to the character array accessed through `c_str()` is undefined behavior.

And the last thing you want is more UB.

Then there's `data()`, too, but this is essentially the same as `c_str()`:

> c_str() and data() perform the same function. (since C++11)

So chances are that is UB, too.

I can understand the desire to get rid of the allocation, but if I have to decide between one more allocation or risking undefined behaviour, then I take the allocation.

Reply:
Ahh the non-const .data() overload is C++17 feature. 
There's always `&str[0]` ;)

Reply:
> Ahh the non-const .data() overload is C++17 feature.

Indeed.

> There's always `&str[0]` ;)

Feels a bit hacky to me, I did it to remove the `alloca()` call.